File: src/Wc3FrameExtLua/Fdf/Text.lua
--=========
-- Include
--=========

local Class = LibManager.getDepency('LuaClass') or error('')
---@type Wc3Utils
local Wc3Utils = LibManager.getDepency('Wc3Utils') or error('')
local isTypeErr = Wc3Utils.isTypeErr or error('')
local Log = Wc3Utils.Log or error('')

---@type FdfDataClass
local FdfData = require('FdfData') or error('')

--=======
-- Class
--=======

local FdfText = Class.new('FdfText', FdfData)
---@class FdfText : FdfData
local public = FdfText.public
---@class FdfTextClass : FdfDataClass
local static = FdfText.static
---@type FdfTextClass
local override = FdfText.override
local private = {}

--=========
-- Static
--=========

---@param name string
---@param child FdfText | nil
---@return FdfText
function override.new(name, child)
    isTypeErr(name, 'string', 'name')
    if child then isTypeErr(child, FdfText, 'child') end

    local instance = child or Class.allocate(FdfText)
    instance = FdfData.new(name, 'TEXT', false, nil, instance)

    return instance
end

--========
-- Public
--========

---@param width number
function public:setWidth(width)
    isTypeErr(self, FdfText, 'self')
    isTypeErr(width, 'number', 'width')

    local str_width = tostring(width)
    self:setParameter('Width', str_width)
end

---@param height number
function public:setHeight(height)
    isTypeErr(self, FdfText, 'self')
    isTypeErr(height, 'number', 'height')

    local str_height = tostring(height)
    self:setParameter('Height', str_height)
end

--- Look-Up FileNames in some String table (for example gameinterface)
---@param flag boolean
function public:setDecorateFileNames(flag)
    isTypeErr(self, FdfText, 'self')
    isTypeErr(flag, 'boolean', 'flag')

    if flag then
        self:setParameter('DecorateFileNames')
    else
        self:removeParameter('DecorateFileNames')
    end
end

---@param text string
function public:setText(text)
    isTypeErr(self, FdfText, 'self')
    isTypeErr(text, 'string', 'text')

    self:setParameter('Text', '\"'..text..'\"')
end

---@param path string
---@param size number
function public:setFont(path, size)
    isTypeErr(self, FdfText, 'self')
    isTypeErr(path, 'string', 'path')
    isTypeErr(size, 'number', 'size')

    self:setParameter('FrameFont', string.format('\"%s\", %f, \"\"',
                                                 path, size))
end

---@param horz string | "'JUSTIFYCENTER'" | "'JUSTIFYLEFT'" | "'JUSTIFYRIGHT'"
---@param vert string | "'JUSTIFYMIDDLE'" | "'JUSTIFYTOP'" | "'JUSTIFYBOTTOM'"
function public:setJustification(horz, vert)
    isTypeErr(self, FdfText, 'self')
    isTypeErr(horz, 'string', 'horz')
    isTypeErr(vert, 'string', 'vert')

    self:setParameter('FontJustificationH', horz)
    self:setParameter('FontJustificationV', vert)
end

---@param x number
---@param y number
function public:setJustificationOffset(x, y)
    isTypeErr(self, FdfText, 'self')
    isTypeErr(x, 'number', 'x')
    isTypeErr(y, 'number', 'y')

    self:setParameter('FontJustificationOffset', tostring(x)..' '..tostring(y))
end

--- [0,1]
---@param r number
---@param g number
---@param b number
---@param a number
function public:setColor(r, g, b, a)
    isTypeErr(self, FdfText, 'self')
    isTypeErr(r, 'number', 'r')
    isTypeErr(g, 'number', 'g')
    isTypeErr(b, 'number', 'b')
    isTypeErr(a, 'number', 'a')

    r = r < 0 and 0 or r > 1 and 1 or r
    g = g < 0 and 0 or g > 1 and 1 or g
    b = b < 0 and 0 or b > 1 and 1 or b
    a = a < 0 and 0 or a > 1 and 1 or a

    self:setParameter('FontColor', string.format('%f %f %f %f',
                                                 r, g, b, a))
end

--- [0,1]
---@param r number
---@param g number
---@param b number
---@param a number
function public:setHighlightColor(r, g, b, a)
    isTypeErr(self, FdfText, 'self')
    isTypeErr(r, 'number', 'r')
    isTypeErr(g, 'number', 'g')
    isTypeErr(b, 'number', 'b')
    isTypeErr(a, 'number', 'a')

    r = r < 0 and 0 or r > 1 and 1 or r
    g = g < 0 and 0 or g > 1 and 1 or g
    b = b < 0 and 0 or b > 1 and 1 or b
    a = a < 0 and 0 or a > 1 and 1 or a

    self:setParameter('FontHighlightColor', string.format('%f %f %f %f',
                                                 r, g, b, a))
end

--- [0,1]
---@param r number
---@param g number
---@param b number
---@param a number
function public:setDisabledColor(r, g, b, a)
    isTypeErr(self, FdfText, 'self')
    isTypeErr(r, 'number', 'r')
    isTypeErr(g, 'number', 'g')
    isTypeErr(b, 'number', 'b')
    isTypeErr(a, 'number', 'a')

    r = r < 0 and 0 or r > 1 and 1 or r
    g = g < 0 and 0 or g > 1 and 1 or g
    b = b < 0 and 0 or b > 1 and 1 or b
    a = a < 0 and 0 or a > 1 and 1 or a

    self:setParameter('FontDisabledColor', string.format('%f %f %f %f',
                                                 r, g, b, a))
end

--- [0,1]
---@param r number
---@param g number
---@param b number
---@param a number
function public:setShadowColor(r, g, b, a)
    isTypeErr(self, FdfText, 'self')
    isTypeErr(r, 'number', 'r')
    isTypeErr(g, 'number', 'g')
    isTypeErr(b, 'number', 'b')
    isTypeErr(a, 'number', 'a')

    r = r < 0 and 0 or r > 1 and 1 or r
    g = g < 0 and 0 or g > 1 and 1 or g
    b = b < 0 and 0 or b > 1 and 1 or b
    a = a < 0 and 0 or a > 1 and 1 or a

    self:setParameter('FontShadowColor', string.format('%f %f %f %f',
                                                 r, g, b, a))
end

---@param x number
---@param y number
function public:setShadowOffset(x, y)
    isTypeErr(self, FdfText, 'self')
    isTypeErr(x, 'number', 'x')
    isTypeErr(y, 'number', 'y')

    self:setParameter('FontShadowOffset', tostring(x)..' '..tostring(y))
end

--=========
-- Private
--=========

return static

File: src/Fdf/Fdf/Text.ts
import { Color } from "../../Utils";
import { Fdf } from "../Fdf";

export class Text extends Fdf {
    constructor(name: string){
        super(name, 'TEXT', false)

        this.__width = -1
        this.__height = -1
        this.__decorate = false
        this.__text = ''
        this.__font = ''
        this.__font_size = 0.01
        this.__just_horz = 'JUSTIFYCENTER'
        this.__just_vert = 'JUSTIFYMIDDLE'
    }

    get width(){return this.__width}
    set width(w: number){
        this._setParam('Width', w.toString())
        this.__width = w
    }

    get height(){return this.__height}
    set height(h: number){
        this._setParam('Height', h.toString())
        this.__height = h
    }

    get decorateFileNames(){return this.__decorate}
    set decorateFileNames(flag: boolean){
        if (flag){
            this._setParam('DecorateFileNames')
        } else {
            this._removeParam('DecorateFileNames')
        }
        this.__decorate = flag
    }

    get text(){return this.__text}
    set text(text: string){
        this._setParam('Text',  '\"' + text + '\"')
        this.__text = text
    }

    get font(){return this.__font}
    set font(path: string){
        this.__font = path
        this._setParam('FrameFont',
                       string.format('\"%s\", %f, \"\"', this.__font, this.__font_size))
    }

    get fontSize(){return this.__font_size}
    set fontSize(size: number){
        this.__font_size = size
        this._setParam('FrameFont',
                       string.format('\"%s\", %f, \"\"', this.__font, this.__font_size))
    }

    get justification(){return [this.__just_horz, this.__just_vert]}
    set justification(val: [FdfText.JustificationHorz, FdfText.JustificationVert]){
        this.__just_horz = val[0]
        this.__just_vert = val[1]
        
        this._setParam('FontJustificationH', this.__just_horz)
        this._setParam('FontJustificationV', this.__just_vert)
    }

    private __width: number 
    private __height: number
    private __decorate: boolean
    private __text: string
    private __font: string
    private __font_size: number
    private __just_horz: FdfText.JustificationHorz
    private __just_vert: FdfText.JustificationVert

    // TODO
    private __just_offset: [number, number] = [0, 0]
    private __color: Color = new Color()
    private __color_highlight: Color = new Color()
    private __color_disabled: Color = new Color()
    private __color_shadow: Color = new Color()
    private __shadow_offset: [number, number] = [0, 0]
}

export namespace FdfText {
    export type JustificationHorz = 'JUSTIFYLEFT'|'JUSTIFYCENTER'|'JUSTIFYRIGHT'
    export type JustificationVert = 'JUSTIFYTOP'|'JUSTIFYMIDDLE'|'JUSTIFYBOTTOM'
}